From 9853adfb71ad0d6077577a2a513e6cc4f19559aa Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Mon, 24 Jul 2023 12:18:07 +0200
Subject: [PATCH] board: cn9131-cf-solidwan: switch cp0_phy0 to
 auto-negotiation

CP0 mdio bus has an address conflict for 1x PHY on the SoM, and 1x PHY
on the SolidWAN Carrier at address 0.

Configure the PHY(s) at boot-time for SGMII to Copper, with
auto-negotiation enabled between PHY and MAC so that MDIO messaging
becomes unnecessary.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 arch/arm/dts/cn9130-cf-solidwan.dts    | 14 +++---
 board/Marvell/octeontx2_cn913x/board.c | 61 ++++++++++++++++++++++++++
 configs/sr_cn913x_cex7_defconfig       |  1 +
 3 files changed, 71 insertions(+), 5 deletions(-)

diff --git a/arch/arm/dts/cn9130-cf-solidwan.dts b/arch/arm/dts/cn9130-cf-solidwan.dts
index 98c874f63b..3154316c87 100644
--- a/arch/arm/dts/cn9130-cf-solidwan.dts
+++ b/arch/arm/dts/cn9130-cf-solidwan.dts
@@ -300,8 +300,12 @@
 
 /* SRDS #3 - 1GE PHY over SGMII */
 &cp0_eth1 {
-	phy = <&cp0_phy0>;
+	/*
+	 * Due to mdio address conflict use SGMII Auto-Negotation
+	 * between MAC and PHY only, skipping MDIO.
+	 */
 	phy-mode = "sgmii";
+	managed = "in-band-status";
 	/*
 	 * All Carrier PHYs share single reset line
 	 * phy-reset-gpios = <&cp0_gpio1 7 GPIO_ACTIVE_LOW>; // MPP39
@@ -342,10 +346,10 @@
 	status = "okay";
 	pinctrl-0 = <&cp0_mdio_pins>;
 
-	cp0_phy0: ethernet-phy@0 {
-		marvell,reg-init = <3 16 0 0x1017>;
-		reg = <0>;
-	};
+	/*
+	 * SoM + Carrier both have PHY at address 0,
+	 * ignore MDIO and use SGMII Auto-Negotation between MAC and PHY.
+	 */
 
 	cp0_phy1: ethernet-phy@1 {
 		marvell,reg-init = <3 16 0 0x1017>;
diff --git a/board/Marvell/octeontx2_cn913x/board.c b/board/Marvell/octeontx2_cn913x/board.c
index 4df0c33b84..2e12072e88 100644
--- a/board/Marvell/octeontx2_cn913x/board.c
+++ b/board/Marvell/octeontx2_cn913x/board.c
@@ -18,6 +18,7 @@
 #ifdef CONFIG_BOARD_CONFIG_EEPROM
 #include <mvebu/cfg_eeprom.h>
 #endif
+#include <miiphy.h>
 
 #define CP_USB20_BASE_REG(cp, p)	(MVEBU_REGS_BASE_CP(0, cp) + \
 						0x00580000 + 0x1000 * (p))
@@ -255,3 +256,63 @@ int board_late_init(void)
 #endif
 	return 0;
 }
+
+/*
+ * Wrapper around miphy_read & miiphy_write to allow modifying by bitmask
+ */
+static inline int miiphy_modify(const char *devname, unsigned char addr, unsigned char reg, unsigned short value, unsigned short mask) {
+	unsigned short old, new;
+
+	if (miiphy_read(devname, addr, reg, &old)) {
+		printf("read failed ...\n");
+		return 1;
+	}
+
+	new = old & (~mask);
+	new |= (value & mask);
+
+	return miiphy_write(devname, addr, reg, new);
+}
+
+/*
+ * CN9131 SolidWAN has a PHY address conflict between SoM and Carrier.
+ *
+ * Configure PHY(s) at address 0 for SGMII auto-negotiation between MAC and PHY,
+ */
+static void solidwan_phy_init() {
+	const char *mii_bus_name = "cp0-mdio@12a200";
+	struct mii_dev *bus;
+	int ret = 0;
+
+	bus = miiphy_get_dev_by_name(mii_bus_name);
+	if (!bus) {
+		printf("Warning: Failed to get \"%s\", can't configure phy for eth1!\n", mii_bus_name);
+		return;
+	}
+
+	ret |= miiphy_write(mii_bus_name, 0x00, 0x16, 0x0001); // select page 1
+	ret |= miiphy_modify(mii_bus_name, 0x00, 0x00, 0x1000, 0x1000); // set page 1 register 0 bit 12=1
+	ret |= miiphy_modify(mii_bus_name, 0x00, 0x1a, 0x0000, 0x0040); // set page 1 register 26 bit 6=0
+	ret |= miiphy_modify(mii_bus_name, 0x00, 0x00, 0x8000, 0x8000); // set page 1 register 0 bit 15=1
+	udelay(1000); // wait for sw-reset to complete
+	ret |= miiphy_write(mii_bus_name, 0x00, 0x16, 0x0012); // select page 18
+	ret |= miiphy_modify(mii_bus_name, 0x00, 0x14, 0x0001, 0x0007); // set page 18 register 20 bits [2:0]=001
+	ret |= miiphy_modify(mii_bus_name, 0x00, 0x14, 0x8000, 0x8000); // set page 18 register 20 bit 15=1
+	udelay(1000); // wait for sw-reset to complete
+
+	if (ret)
+		printf("Warning: Failed to configure phy for eth1!\n");
+}
+
+int last_stage_init(void)
+{
+	const char *fdtfile;
+
+	fdtfile = env_get("fdtfile");
+	if (strcmp(fdtfile, "marvell/cn9131-cf-solidwan.dtb") == 0) {
+		printf("Applying eth1 phy workaround ...\n");
+		solidwan_phy_init();
+	}
+
+	return 0;
+}
diff --git a/configs/sr_cn913x_cex7_defconfig b/configs/sr_cn913x_cex7_defconfig
index d7445e9476..33c16d7412 100644
--- a/configs/sr_cn913x_cex7_defconfig
+++ b/configs/sr_cn913x_cex7_defconfig
@@ -105,3 +105,4 @@ CONFIG_I2C_EEPROM=y
 CONFIG_CMD_TLV_EEPROM=y
 CONFIG_SUPPORT_EMMC_BOOT=y
 CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_LAST_STAGE_INIT=y
-- 
2.35.3

